lib/puppet/provider/openldap_schema/olc.rb

```diff
@@ -0,0 +1,123 @@
+require 'tempfile'
+
+Puppet::Type.type(:openldap_schema).provide(:olc) do
+
+  # TODO: Use ruby bindings (can't find one that support IPC)
+
+  defaultfor :osfamily => :debian, :osfamily => :redhat
+
+  commands :slapcat => 'slapcat', :ldapadd => 'ldapadd', :slaptest => 'slaptest', :ldapsearch => 'ldapsearch'
+
+  mk_resource_methods
+
+  @@openldap_schema_existing = []
+  @@openldap_schema_new = []
+  @@openldap_schema_create_count = 0
+  @@openldap_schema_max_count = 0
+
+  def self.instances
+    self.getSchemaNames.map { |schema|
+          new(
+            :ensure => :present,
+            :name   => schema
+          )
+    }
+  end
+
+  def self.getSchemaNames
+      schemas = []
+      ldapsearch('-Y', 'EXTERNAL', '-H', 'ldapi:///', '-b', 'cn=config', '(objectClass=olcSchemaConfig)', 'cn').split("\n\n").each do |paragraph|
```

Comment:
I switched the slapcat and slapadd commands with the ldapadd and ldapsearch commands because slapadd cannot run while slapd is running. I'll try rewriting it to use slapcat again.